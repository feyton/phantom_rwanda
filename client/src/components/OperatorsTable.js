import React from 'react';
import { Link } from 'react-router-dom';
import { useGlobalFilter, usePagination, useTable } from 'react-table';
import { useSelector } from 'react-redux';
import CheckRole from '../utils/CheckRoles.js';
import Button from './Button.js';
import ManageDropDownOperator from './ManageDropDownOperator.js';
import SearchFilter from './SearchFilter.js';
import { TableColumns } from './tableColumns.js';

const OperatorsTable = ({ data }) => {
  const { user, authenticated } = useSelector((state) => state?.auth);
  const manageCellValue = ({ row }) => (
    <ManageDropDownOperator user={user} row={row} />
  );
  const columns = TableColumns(
    'Mobile Number',
    'mobileNumber',
    'Company',
    'company',
    manageCellValue
  );
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    page,
    nextPage,
    previousPage,
    canPreviousPage,
    canNextPage,
    gotoPage,
    pageCount,
    setPageSize,
    prepareRow,
    pageOptions,
    state,
    setGlobalFilter
  } = useTable(
    {
      columns,
      data
    },
    useGlobalFilter,
    usePagination
  );

  const { globalFilter, pageIndex, pageSize } = state;

  return (
    <>
      <section className="flex justify-between content-center pb-7">
        <SearchFilter
          filter={globalFilter}
          setfilter={setGlobalFilter}
          placeholder="Search Operators..."
        />
        <CheckRole
          children={
            <Link to="operator/register">
              <Button
                name="Register new"
                styles="bg-primary text-white py-1 text-xs"
              />
            </Link>
          }
          role={['admin']}
        />
      </section>
      <table
        {...getTableProps()}
        className="w-full overflow-x-scroll font-raleway"
      >
        <thead>
          {headerGroups.map((headerGroup) => (
            <tr {...headerGroup.getHeaderGroupProps()} className="pb-10 ">
              {headerGroup.headers.map((column) => (
                <th
                  {...column.getHeaderProps()}
                  className="text-left font-extrabold pr-5 border-b-2 pb-4"
                >
                  {column.render('Header')}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {page.map((row) => {
            prepareRow(row);
            return (
              <tr {...row.getRowProps()}>
                {row.cells.map((cell) => {
                  return (
                    <td
                      {...cell.getCellProps()}
                      className="pr-5 py-4 border-b-2"
                    >
                      {cell.render('Cell')}
                    </td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
      {pageOptions.length > 1 && (
        <table className=" mt-3 flex justify-center w-full bg-gray-300 bg-opacity-20 ">
          <tfoot>
            <tr className="p-1 flex-row flex items-center justify-center my-2 mx-auto w-full font-raleway">
              <td colSpan={5}>
                <div className="w-full justify-center flex mx-auto flex-row items-center overflow-x-scroll">
                  <button
                    className="mx-2 rounded-circle font-bold flex items-center justify-center bg-primary h-[30px] w-[60px] cursor-pointer"
                    onClick={() => gotoPage(0)}
                    disabled={!canPreviousPage}
                  >
                    {'<|'}
                  </button>{' '}
                  <button
                    className="mx-2 font-bold font-raleway"
                    onClick={() => previousPage()}
                    disabled={!canPreviousPage}
                  >
                    Prev
                  </button>{' '}
                  <button
                    onClick={() => nextPage()}
                    disabled={!canNextPage}
                    className="mx-2 font-bold font-raleway cursor-pointer"
                  >
                    Next
                  </button>{' '}
                  <button
                    onClick={() => gotoPage(pageCount - 1)}
                    disabled={!canNextPage}
                    className="mx-2 rounded-circle font-bold flex items-center justify-center bg-primary h-[30px] w-[60px] cursor-pointer"
                  >
                    {'|>'}
                  </button>{' '}
                  <div className="flex flex-row justify-center w-full">
                    <span className="inline-block mx-2">
                      Page{' '}
                      <strong>
                        {pageIndex + 1} of {pageOptions.length}
                      </strong>{' '}
                    </span>
                    <span className="inline-block mx-2">
                      | Go to page:{' '}
                      <input
                        type="number"
                        className=" pl-2 outline-none border rounded-md border-primary"
                        defaultValue={pageIndex + 1}
                        onChange={(e) => {
                          const pageNumber = e.target.value
                            ? Number(e.target.value) - 1
                            : 0;
                          gotoPage(pageNumber);
                        }}
                        style={{ width: '50px' }}
                      />
                    </span>{' '}
                    <select
                      className="px-1/2 md:px-2 font-raleway rounded-md border border-primary"
                      value={pageSize}
                      onChange={(e) => setPageSize(Number(e.target.value))}
                    >
                      {[10, 25, 50].map((pageSize) => (
                        <option key={pageSize} value={pageSize}>
                          Show {pageSize}
                        </option>
                      ))}
                    </select>
                  </div>
                </div>
              </td>
            </tr>
          </tfoot>
        </table>
      )}
    </>
  );
};

export default OperatorsTable;
