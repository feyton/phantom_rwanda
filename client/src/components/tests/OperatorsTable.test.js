import React from 'react';
import { render } from '@testing-library/react';
import OperatorsTable from '@components/OperatorsTable.js';
import ProviderWrapper from '@utils/TestUtil.js';

const data = [
  {
    address: 'Kabuga, Kigali',
    company: 'Kigali Bus Services',
    id: 2,
    license: '1200080081693164',
    mobileNumber: '0788352746',
    nationalID: '1200080081391164',
    user: {
      email: 'feyton@hotmail.com',
      firstName: 'testName',
      id: 13,
      lastName: 'testname2',
      role: 'driver'
    }
  },
  {
    address: 'Kabuga, Kigali',
    company: 'Kigali Bus Services',
    id: 2,
    license: '1200080081693164',
    mobileNumber: '0785552746',
    nationalID: '1200080085591164',
    user: {
      email: 'driver@hotmail.com',
      firstName: 'feyton',
      id: 13,
      lastName: 'testname2',
      role: 'driver'
    }
  }
];

describe('Operator Table', () => {
  it('Renders the table', () => {
    const elem = render(
      <ProviderWrapper>
        <OperatorsTable data={data} />
      </ProviderWrapper>
    );
    expect(elem).toMatchSnapshot();
  });
});
