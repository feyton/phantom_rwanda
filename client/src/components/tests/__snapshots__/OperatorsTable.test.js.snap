// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operator Table Renders the table 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="flex justify-between content-center pb-7"
      >
        <div>
          <input
            class="p-1 rounded-md border-2"
            name="search"
            placeholder="Search Operators..."
            type="text"
            value=""
          />
        </div>
        
      </section>
      <table
        class="w-full overflow-x-scroll font-raleway"
        role="table"
      >
        <thead>
          <tr
            class="pb-10 "
            role="row"
          >
            <th
              class="text-left font-extrabold pr-5 border-b-2 pb-4"
              colspan="1"
              role="columnheader"
            >
              Names
            </th>
            <th
              class="text-left font-extrabold pr-5 border-b-2 pb-4"
              colspan="1"
              role="columnheader"
            >
              Email
            </th>
            <th
              class="text-left font-extrabold pr-5 border-b-2 pb-4"
              colspan="1"
              role="columnheader"
            >
              Mobile Number
            </th>
            <th
              class="text-left font-extrabold pr-5 border-b-2 pb-4"
              colspan="1"
              role="columnheader"
            >
              Company
            </th>
            <th
              class="text-left font-extrabold pr-5 border-b-2 pb-4"
              colspan="1"
              role="columnheader"
            >
              Management
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              testName
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              feyton@hotmail.com
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              0788352746
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              Kigali Bus Services
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              <div
                class="z-10 font-sm"
              >
                <select
                  class="py-1 px-1 font-rale font-bold bg-transparent border rounded-sm"
                  id="select"
                  name=""
                  role="select"
                >
                  <option
                    hidden=""
                    value="manage"
                  >
                    Manage
                  </option>
                  <option
                    value="view"
                  >
                    View
                  </option>
                  
                  
                  
                </select>
              </div>
            </td>
          </tr>
          <tr
            role="row"
          >
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              feyton
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              driver@hotmail.com
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              0785552746
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              Kigali Bus Services
            </td>
            <td
              class="pr-5 py-4 border-b-2"
              role="cell"
            >
              <div
                class="z-10 font-sm"
              >
                <select
                  class="py-1 px-1 font-rale font-bold bg-transparent border rounded-sm"
                  id="select"
                  name=""
                  role="select"
                >
                  <option
                    hidden=""
                    value="manage"
                  >
                    Manage
                  </option>
                  <option
                    value="view"
                  >
                    View
                  </option>
                  
                  
                  
                </select>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <section
      class="flex justify-between content-center pb-7"
    >
      <div>
        <input
          class="p-1 rounded-md border-2"
          name="search"
          placeholder="Search Operators..."
          type="text"
          value=""
        />
      </div>
      
    </section>
    <table
      class="w-full overflow-x-scroll font-raleway"
      role="table"
    >
      <thead>
        <tr
          class="pb-10 "
          role="row"
        >
          <th
            class="text-left font-extrabold pr-5 border-b-2 pb-4"
            colspan="1"
            role="columnheader"
          >
            Names
          </th>
          <th
            class="text-left font-extrabold pr-5 border-b-2 pb-4"
            colspan="1"
            role="columnheader"
          >
            Email
          </th>
          <th
            class="text-left font-extrabold pr-5 border-b-2 pb-4"
            colspan="1"
            role="columnheader"
          >
            Mobile Number
          </th>
          <th
            class="text-left font-extrabold pr-5 border-b-2 pb-4"
            colspan="1"
            role="columnheader"
          >
            Company
          </th>
          <th
            class="text-left font-extrabold pr-5 border-b-2 pb-4"
            colspan="1"
            role="columnheader"
          >
            Management
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            testName
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            feyton@hotmail.com
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            0788352746
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            Kigali Bus Services
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            <div
              class="z-10 font-sm"
            >
              <select
                class="py-1 px-1 font-rale font-bold bg-transparent border rounded-sm"
                id="select"
                name=""
                role="select"
              >
                <option
                  hidden=""
                  value="manage"
                >
                  Manage
                </option>
                <option
                  value="view"
                >
                  View
                </option>
                
                
                
              </select>
            </div>
          </td>
        </tr>
        <tr
          role="row"
        >
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            feyton
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            driver@hotmail.com
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            0785552746
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            Kigali Bus Services
          </td>
          <td
            class="pr-5 py-4 border-b-2"
            role="cell"
          >
            <div
              class="z-10 font-sm"
            >
              <select
                class="py-1 px-1 font-rale font-bold bg-transparent border rounded-sm"
                id="select"
                name=""
                role="select"
              >
                <option
                  hidden=""
                  value="manage"
                >
                  Manage
                </option>
                <option
                  value="view"
                >
                  View
                </option>
                
                
                
              </select>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
